### DISCUSSION TITLE
Would a port of dataclasses to micropython-lib be welcome?

### DISCUSSION THREAD

**Discussion by winni2k** [micropython/micropython#13741] [Core Development]
I think dataclasses are the best. I would be interested in trying my hand at porting the module to micropython-lib. Is that a terrible idea or a great idea?

**Comment by davefes** [comment]
Rewriting some of my applications using dataclass was my next learning exercise.  I had found this:
https://github.com/orgs/micropython/discussions/9327
which seemed to suggest that it was possible.

Have I misunderstood?

**Reply by winni2k** [reply]
Right! I think I missed a few of the earlier comments when I first read that post and it made no sense to me. Now on a second read, it appears as if there is an implementation of dataclasses, but it has not been shared with micropython-lib?

**Reply by davefes** [reply]
I will await further developments.

**Reply by winni2k** [reply]
It looks like there is no implementation available from that discussion (https://github.com/orgs/micropython/discussions/9327#discussioncomment-8635740).

**Comment by mattytrentini** [comment]
Dataclasses are desirable and would certainly be a welcome feature if they could be implemented with reasonable resource usage - and high compatibility with CPython's implementation.

**Reply by winni2k** [reply]
I'll see if I can have a go then. However, I am new to micropython (not python). Can you give me a pointer or two on how to measure relevant resource usage?

**Reply by winni2k** [reply]
I have started by porting a pre-requisite, `functools.total_ordering` (https://github.com/micropython/micropython-lib/pull/818). I could use a PR review.

Out of the gate, it appears as if `dataclasses` will not work if they are used to extend, that is, inherit from builtin types because builtin types do not appear to have comparison operators as attributes (for example `int.__lt__()`).

**Comment by ambv** [comment]
There's a dataclasses-like library that works on MicroPython:
https://udataclasses.readthedocs.io/en/latest/

**Reply by davefes** [reply]
Could you provide a link to the library?  Thank you.

**Reply by AndrewD** [reply]
This looks like it: 
[udataclasses](https://github.com/dhrosa/udataclasses)

**Reply by davefes** [reply]
Thank you. Looks like the keyword was `dhrosa` as `udataclasses` didn't turn-up anything other than the posted link.
